third-party/rsutils/src/adapter-watcher.cpp

```diff
@@ -0,0 +1,236 @@
+
+                  CancelIPChangeNotify( &overlap );
+              } )
+#else
+        // https://stackoverflow.com/questions/27008067/how-to-get-notified-about-network-interface-changes-with-netlist-and-rtmgrp-link
+        : _socket( open_netlink() )
+        , _done( eventfd( 0, 0 ) )
+        , _th(
+              [this]
+              {
+                  // We're going to wait on the following handles:
+                  LOG_DEBUG( "starting network adapter watcher" );
+
+                  int const timeout = -1;  // negative value = infinite
+                  int const nfds = 2;
+                  pollfd fds[nfds];
+                  memset( fds, 0, sizeof( fds ) );
+                  fds[0].fd = _done;
+                  fds[0].events = POLLIN;
+                  fds[1].fd = _socket;
+                  fds[1].events = POLLIN;
+
+                  while( true )
+                  {
+                      fds[0].revents = 0;
+                      fds[1].revents = 0;
+                      auto rv = poll( fds, nfds, timeout );
+                      if( rv <= 0 )
+                      {
+                          LOG_ERROR( "poll failed: " << rv );
+                          break;
+                      }
+
+                      if( fds[0].revents != 0 )
+                          break;  // we're done
+
+                      LOG_DEBUG( "network adapter changes detected!" );
+
+                      char buf[4096];
+                      struct sockaddr_nl snl;
+                      iovec iov = { buf, sizeof buf };
+                      msghdr msg = { (void *)&snl, sizeof snl, &iov, 1, NULL, 0, 0 };
+                      int status = recvmsg( _socket, &msg, 0 );
+                      if( status < 0 )
+                      {
+                          // Socket non-blocking so bail out once we have read everything
+                          if( errno == EWOULDBLOCK || errno == EAGAIN )
+                              continue;
+
+                          // Anything else is an error
+                          LOG_DEBUG( "read_netlink: Error recvmsg: " << status );
+                          break;
+                      }
+                      if( status == 0 )
+                          LOG_DEBUG( "read_netlink: EOF" );
+
+                      // We need to handle more than one message per recvmsg
+                      for( auto h = (struct nlmsghdr *)buf; NLMSG_OK( h, (unsigned int)status );
+                           h = NLMSG_NEXT( h, status ) )
+                      {
+                          // Finish reading
+                          if( h->nlmsg_type == NLMSG_DONE )
+                              continue;
+
+                          // Message is some kind of error
+                          if( h->nlmsg_type == NLMSG_ERROR )
+                          {
+                              LOG_DEBUG( "read_netlink: Message is an error - decode TBD" );
```

Comment:
LOG_ERROR / WARNING??

Reply:
Worst case, we do not get adapter change notifications, so no: I don't think it should be a warning or error.

Reply:
I made it a generic "read_netlink: got NLMSG_ERROR" message, still debug: I don't want to start flooding with warnings or errors since we do not really exit the loop. We just continue going over all the messages.